### ISSUE TITLE
CLI has several bugs relating to SAML SSO

### ISSUE THREAD

**Issue by roffelsaurus** [sanity-io/sanity#2993]
**Describe the bug**

Prereq: Business tier SAML is set up normally and functions just fine in both studio, cli and management dashboard.

1. You login with `sanity login --sso slug`, then wait a day.

Bug/poor DX 1:

2a. You attempt any CLI command that uses the api and get `Session is expired`. This is fine in the browser where you would expect this to lead a redirect train back inside the walls( However, that also seems to be bugged and possibly related to this: #2985 ). In the CLI it is way too involved to login again every day. It also differs from how sanity CLI has worked for us the last 2.5 years for both user+pass login and github SSO, neither of which required relogin in CLI.

Bug/horrible DX 2:

2b: You attempt `sanity logout` and get `session is expired`. This is even worse DX, as you can go precisely nowhere without hacking on the issue. If the token is stale and you can't logout remotely, at least delete the token clientside.

Bug/horrible DX3:

2b: You attempt `sanity login --sso slug` and get `session is expired`. This is pretty much the same problem as 2b, where you can't escape the stale session any way.

**Workaround**

Deleting `/user/.config/sanity/config.json` purges the token and you can login again. This hints at how the CLI can improve DX as well.

**Expected behavior**

Most importantly, `sanity logout` should purge stale sessions.

Additionally, `sanity login` should not crash with session is expired. The user wants to login. Debugging it I see that it is because it tries to fetch providers, but that should not be necessary when specifying provider and as a fallback it definitely should read default-login.json, not crash.

Thirdly, SAML SSO token session in cli should get a robot token like expiry time, like months and years, not 1 day like a browser pleb.

**Which versions of Sanity are you using?**

@sanity/cli             2.22.3 (latest: 2.23.0)
@sanity/base            2.22.0 (latest: 2.23.0)
@sanity/cli             2.22.0 (latest: 2.23.0)
@sanity/components      2.14.0 (up to date)
@sanity/core            2.22.0 (latest: 2.23.0)
@sanity/default-layout  2.22.0 (latest: 2.23.0)
@sanity/default-login   2.22.0 (latest: 2.23.0)
@sanity/desk-tool       2.22.2 (latest: 2.23.0)
@sanity/vision          2.22.0 (latest: 2.23.0)

**What operating system are you using?**

**Which versions of Node.js / npm are you running?**

6.14.14
v14.17.5

**Comment by rexxars** [comment]
Thank you very much for filing this @roffelsaurus - this is definitely highlight a number of flaws in the way the CLI handles login/logout.

I've opened #3001 and #3002 for addressing some of these issues, and will discuss with the team how we approach some of the other issues you've brought up.

**Comment by roffelsaurus** [comment]
Thanks for the quick response and fixes. That'll fix the CLI DX considerably. 

I found a compromise regarding my last unaddressed point which would have been an issue for your backenders;

> Thirdly, SAML SSO token session in cli should get a robot token like expiry time, like months and years, not 1 day like a browser pleb.

I found that we can likely reduce CLI reauthentications by changing the SAML session expiry time from 12 hours to something a bit more in the SAML config at manage.sanity.io. This will affect browsers too, but that is acceptable
